charms/worker/k8s/src/pki.py

```diff
@@ -6,57 +6,49 @@
 
 import logging
-import subprocess
-from ipaddress import ip_address
 from pathlib import Path
 from typing import List, Tuple, Union
 
+from cryptography import x509
+
 log = logging.getLogger(__name__)
 
 
 def get_certificate_sans(cert_path: Union[str, Path]) -> Tuple[List[str], List[str]]:
     """Extract the DNS and IP Subject Alternative Names (SANs) from a given certificate file.
 
-    This function uses the openssl command to extract the SANs from the certificate file.
+    This function uses the cryptography library to extract the SANs from the certificate file.
 
     Args:
         cert_path (Union[str, Path]): The path to the certificate file.
 
     Returns:
-        tuple[list[str], list[str]]: A tuple containing two lists:
+        Tuple[List[str], List[str]]: A tuple containing two lists:
             - The first list contains DNS SANs.
             - The second list contains IP SANs.
     """
-    try:
-        cmd = f"openssl x509 -noout -ext subjectAltName -in {cert_path}".split()
-        result = subprocess.run(cmd, capture_output=True, text=True, check=True)
-    except subprocess.CalledProcessError as e:
-        log.error("Failed to call openssl for certificate SANs: %s", e)
+    cert_path = Path(cert_path) if isinstance(cert_path, str) else cert_path
+
+    if not cert_path.exists():
+        log.warning("Certificate file not found: %s", cert_path)
         return [], []
 
-    lines = result.stdout.splitlines()
-    if len(lines) < 2:
-        log.info("No SANs found in %s", cert_path)
+    with open(cert_path, "rb") as file:
+        cert_data = file.read()
+
+    try:
+        certificate = x509.load_pem_x509_certificate(cert_data)
+
+        san_extension = certificate.extensions.get_extension_for_class(x509.SubjectAlternativeName)
+        sans = san_extension.value
+
+        dns_sans = set(sans.get_values_for_type(x509.DNSName))
+        ip_address = {str(ip) for ip in sans.get_values_for_type(x509.IPAddress)}
+    except ValueError:
+        log.exception("Failed to parse certificate %s", cert_path)
+        return [], []
+    except x509.ExtensionNotFound:
+        log.exception("No SAN extension found in certificate %s", cert_path)
         return [], []
 
-    # lines[0] == "X509v3 Subject Alternative Name: "
-    all_sans = [san.strip() for san in lines[1].split(",")]
-    dns_sans = set()
-    ip_sans = set()
-
-    dns_prefix = "DNS:"
-    ip_prefix = "IP Address:"
-    for san in all_sans:
-        # fmt: off
-        if san.startswith(dns_prefix):
-            dns_sans.add(san[len(dns_prefix):])
-        elif san.startswith(ip_prefix):
-            ip_str = san[len(ip_prefix):]
-            try:
-                ip = ip_address(ip_str)
-                ip_sans.add(str(ip))
-            except ValueError:
-                log.warning("Invalid IP SAN: %s", ip_str)
-        # fmt: on
-
-    return list(dns_sans), list(ip_sans)
+    return list(dns_sans), list(ip_address)
```

Comment:
How about we sort them.  Is that okay?

```suggestion
    return sorted(dns_sans), sorted(ip_address)
```

Reply:
Sure!